File: 27. Remove Element.cpp
class Solution {
public:
    // Two pointers: O(N), 19.31%, ok
    int removeElement(vector<int>& nums, int val) {
        if(nums.empty()) return 0;
        int head = 0, tail = nums.size() - 1;
        int count = 0;
        while(head <= tail){
            while(head <= tail && nums[head] != val) head++;
            while(head <= tail && nums[tail] == val) {tail--; count++;};
            if(head < tail) {
                swap(nums[head++], nums[tail--]);
                count++;
            }
        }
        nums.resize(nums.size() - count);
        return nums.size();
    }

    // Two pointers, more elegant: O(N), 19.31%, ok
    int removeElement(vector<int>& nums, int val) {
        if(nums.empty()) return 0;
        int valid = 0;
        for(int i = 0; i < nums.size(); i++){
            if(nums[i] != val){
                nums[valid] = nums[i];
                valid++;
            }
        }
        return valid;
    }

    // Two pointers, when elements to remove are rare, avoid unnecessary copy: O(N), 19.31%, ok
    int removeElement(int[] nums, int val) {
        int i = 0;
        int n = nums.length;
        while (i < n) {
            if (nums[i] == val) {
                nums[i] = nums[n - 1];
                n--; // reduce array size by one
            }
            else i++;
        }
        return n;
    }
};

File: 27. Remove Element.py
class Solution(object):
    
    # Two pointers1 (may include unnecessary copy): O(N) time, O(1) space
    def removeElement(self, nums, val):
        i = 0
        for j in xrange(len(nums)):
            if nums[j] != val:
                nums[i] = nums[j]
                i += 1
        return i 
    
    # Two pointers2 (sub-optimal): O(N) time, O(1) space
    def removeElement(self, nums, val):
        i, n = 0, len(nums)
        while i < n:
            if nums[i] == val:
                nums[i] = nums[n - 1]
                n -= 1   # replace last one and decrease length by 1
            else:
                i += 1
        return i 
    
    # Two pointers3 (optimal): O(N) time, O(1) space
    def removeElement(self, nums, val):
        """
        :type nums: List[int]
        :type val: int
        :rtype: int
        """
        i, j = 0, len(nums) - 1
        while i <= j:
            while i <= j and nums[i] != val:
                i += 1
            while i <= j and nums[j] == val:
                j -= 1
            if i <= j:
                nums[i] = nums[j]
                i, j = i + 1, j - 1
        return i 
                
                
        